index.js

```diff
@@ -1,389 +1,156 @@
+ *
+ * @param {String} code Source code
+ * @param {Object} options Parser options
+ * @returns {Object} ESLint parser result object
+ */
+exports.parseForESLint = function parseForESLint(code, options) {
+  var parserOptions = ensureDefaultParserOptions(options);
+  setESLintOptions(parserOptions);
 
-  function checkIdentifierOrVisit(node) {
-    if (node.typeAnnotation) {
-      visitTypeAnnotation.call(this, node.typeAnnotation);
-    } else if (node.type === "Identifier") {
-      this.visit(node);
-    } else {
-      visitTypeAnnotation.call(this, node);
-    }
+  if (!hasPatched) {
+    hasPatched = true;
   }
 
-  function nestTypeParamScope(manager, node) {
-    var parentScope = manager.__currentScope;
-    var scope = new escope.Scope(
-      manager,
-      "type-parameters",
-      parentScope,
-      node,
-      false
-    );
-    manager.__nestScope(scope);
-    for (var j = 0; j < node.typeParameters.params.length; j++) {
-      var name = node.typeParameters.params[j];
-      scope.__define(name, new Definition("TypeParameter", name, name));
-      if (name.typeAnnotation) {
-        checkIdentifierOrVisit.call(this, name);
-      }
-    }
-    scope.__define = function() {
-      return parentScope.__define.apply(parentScope, arguments);
+  function analyzeScope(ast) {
+    var eslintScopeOptions = {
+      ignoreEval: true,
+      impliedStrict: eslintOptions.impliedStrict,
+      sourceType: eslintOptions.sourceType,
+      ecmaVersion: eslintOptions.ecmaVersion,
+      childVisitorKeys: extendedVisitorKeys,
+      fallback: Traverser.getKeys,
     };
-    return scope;
-  }
-
-  // visit decorators that are in: ClassDeclaration / ClassExpression
-  var visitClass = referencer.prototype.visitClass;
-  referencer.prototype.visitClass = function(node) {
-    visitDecorators.call(this, node);
-    var typeParamScope;
-    if (node.typeParameters) {
-      typeParamScope = nestTypeParamScope.call(this, this.scopeManager, node);
-    }
-    // visit flow type: ClassImplements
-    if (node.implements) {
-      for (var i = 0; i < node.implements.length; i++) {
-        checkIdentifierOrVisit.call(this, node.implements[i]);
-      }
-    }
-    if (node.superTypeParameters) {
-      for (var k = 0; k < node.superTypeParameters.params.length; k++) {
-        checkIdentifierOrVisit.call(this, node.superTypeParameters.params[k]);
-      }
-    }
-    visitClass.call(this, node);
-    if (typeParamScope) {
-      this.close(node);
-    }
-  };
 
-  // visit decorators that are in: Property / MethodDefinition
-  var visitProperty = referencer.prototype.visitProperty;
-  referencer.prototype.visitProperty = function(node) {
-    if (node.value && node.value.type === "TypeCastExpression") {
-      visitTypeAnnotation.call(this, node.value);
+    if (eslintOptions.globalReturn !== undefined) {
+      eslintOptions.nodejsScope = eslintOptions.globalReturn;
```

Comment:
FYI, the `eslintOptions.ecmaFeatures.globalReturn` is set by `env: {node: true}`: https://github.com/babel/babel-eslint/blob/42d0c5b20cf65ae4100d06aa8dcaf372723f378d/test/non-regression.js#L17